go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -0,0 +1,293 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vreplication
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"golang.org/x/net/context"
+
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/binlog/binlogplayer"
+	"vitess.io/vitess/go/vt/grpcclient"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vttablet/tabletconn"
+
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+type vcopier struct {
+	vr        *vreplicator
+	tablePlan *TablePlan
+}
+
+func newVCopier(vr *vreplicator) *vcopier {
+	return &vcopier{
+		vr: vr,
+	}
+}
+
+func (vc *vcopier) initTablesForCopy(ctx context.Context) error {
+	defer vc.vr.dbClient.Rollback()
+
+	plan, err := buildReplicatorPlan(vc.vr.source.Filter, vc.vr.tableKeys, nil)
+	if err != nil {
+		return err
+	}
+	if err := vc.vr.dbClient.Begin(); err != nil {
+		return err
+	}
+	// Insert the table list only if at least one table matches.
+	if len(plan.TargetTables) != 0 {
+		var buf strings.Builder
+		buf.WriteString("insert into _vt.copy_state(vrepl_id, table_name) values ")
+		prefix := ""
+		for name := range plan.TargetTables {
+			fmt.Fprintf(&buf, "%s(%d, %s)", prefix, vc.vr.id, encodeString(name))
+			prefix = ", "
+		}
+		if _, err := vc.vr.dbClient.ExecuteFetch(buf.String(), 1); err != nil {
+			return err
+		}
+		if err := vc.vr.setState(binlogplayer.VReplicationCopying, ""); err != nil {
+			return err
+		}
+	} else {
+		if err := vc.vr.setState(binlogplayer.BlpStopped, "There is nothing to replicate"); err != nil {
+			return err
+		}
+	}
+	return vc.vr.dbClient.Commit()
+}
+
+func (vc *vcopier) copyNext(ctx context.Context, settings binlogplayer.VRSettings) error {
+	qr, err := vc.vr.dbClient.ExecuteFetch(fmt.Sprintf("select table_name, lastpk from _vt.copy_state where vrepl_id=%d", vc.vr.id), 10000)
```

Comment:
Random observation: although these db queries are likely to be fast in practice, someday it would be nice if the context propagated to help with timing out / aborting vreplication-related db access.